ARROW-17980: [C++] As-of-Join Substrait extension #14385

westonpace opened the pull request.

rtpsw committed:
ARROW-17980: [C++] As-of-Join Substrait extension

rtpsw committed:
add missing file

rtpsw committed:
add missing proto

rtpsw committed:
CI fixes

rtpsw committed:
distinct keys per input table

rtpsw committed:
CI fixes

rtpsw committed:
resolve conflict

rtpsw committed:
Merge branch 'master' into ARROW-17980

rtpsw committed:
fix typo

rtpsw committed:
Merge branch 'ARROW-17980' of https://github.com/rtpsw/arrow into ARROW-17980

rtpsw committed:
Merge branch 'master' into ARROW-17980

nealrichardson committed:
ARROW-17849: [R][Docs] Document changes due to C++17 for centos-7 users (#14440)

* `configure` checks that `R CMD config CXX17` is defined and exits early if not
* README and install.Rmd discuss system dependencies; install.Rmd includes a bash script to run once to configure your CentOS 7 machine to install arrow. Confirmed that this works manually via docker.
* Updated `r_docker_configure.sh` to use that script logic, so our CI should confirm too.
* Various cleanups in nixlibs.R: remove checks for gcc 4.8/4.9 all over, updated system requirements check to handle case of devtoolset-8 but no openssl/curl
* Unrelated purge of windows build script handling for rtools35

Lead-authored-by: Neal Richardson <neal.p.richardson@gmail.com>
Co-authored-by: Jacob Wujciak-Jens <jacob@wujciak.de>
Signed-off-by: Neal Richardson <neal.p.richardson@gmail.com>

westonpace commented on cpp/cmake_modules/ThirdpartyToolchain.cmake:
```diff
@@ -1767,6 +1771,29 @@ macro(build_substrait)
 
     list(APPEND SUBSTRAIT_SOURCES "${SUBSTRAIT_PROTO_GEN}.cc")
   endforeach()
+  message("SOURCE DIR2 IS ${SOURCE_DIR} AND ${CMAKE_SOURCE_DIR} AND ${ARROW_SUBSTRAIT_PROTOS_DIR}"
+  )
```
Let's remove this cleanup if its ready

rtpsw replied:
Will do.

westonpace commented on cpp/src/arrow/compute/exec/options.h:
```diff
@@ -393,22 +393,35 @@ class ARROW_EXPORT HashJoinNodeOptions : public ExecNodeOptions {
+  /// The keys must be consistent across the input tables:
+  /// Each "on" key must refer to a field of the same type and units across the tables.
+  /// Each "by" key must refer to a list of fields of the same types across the tables.
+  struct Keys {
```
So there is one `Keys` element per input?  Do they all need to have the exact same number of `by_key` keys?

rtpsw replied:
Yes and yes.

westonpace commented on cpp/src/arrow/engine/substrait/options.cc:
```diff
@@ -0,0 +1,111 @@
+  }
+
+ private:
+  Result<DeclarationInfo> MakeAsOfJoinRel(
```
It's a bit weird to have this in `options.cc`.  I think, when we looking at merging into arrow, we will want to find a better home for this.  Perhaps a `custom_relations.cc` or something like that.

rtpsw replied:
Will do.

westonpace commented on cpp/src/arrow/engine/substrait/serde_test.cc:
```diff
@@ -3279,6 +3299,195 @@ TEST(Substrait, IsthmusPlan) {
                        *compute::default_exec_context(), buf, {}, conversion_options);
 }
 
+TEST(Substrait, PlanWithExtension) {
```
Can we change the name of this test to reflect that it has something to do with asof join?

rtpsw replied:
Will do.

westonpace reviewed (commented):
This looks pretty straightforward.  I'll review the substrait changes more closely when we can figure out how to get rid of the `::substrait` change (I think Anja was making progress on this)?

It looks like the major changes are:

 * Change as-of join node so that it can have unique key indices per table
 * Pull the MakeOutputSchema method out into its own header file so it can be referenced by the substrait mapping
 * Add substrait mapping for asof join node

Is that correct?

@rtpsw

raulcd committed:
ARROW-17953: [Archery] Add archery docker info command (#14345)

Some usage examples:

```
$ archery docker info debian-go-cgo --show command
Service debian-go-cgo docker compose config:
 - command: /bin/bash -c "
  /arrow/ci/scripts/go_build.sh /arrow &&
  /arrow/ci/scripts/go_test.sh /arrow"
```
or
```
$ archery docker info debian-go-cgo
Service debian-go-cgo docker compose config:
 - image: ${REPO}:${ARCH}-debian-${DEBIAN}-go-${GO}-cgo
 - build
  - context: .
  - dockerfile: ci/docker/debian-go-cgo.dockerfile
  - cache_from: ['${REPO}:${ARCH}-debian-${DEBIAN}-go-${GO}-cgo']
  - args
   - base: ${REPO}:${ARCH}-debian-${DEBIAN}-go-${GO}
 - shm_size: 2G
 - volumes: ['.:/arrow:delegated', '${DOCKER_VOLUME_PREFIX}debian-ccache:/ccache:delegated']
 - environment
  - ARROW_GO_TESTCGO: 1
 - command: /bin/bash -c "
  /arrow/ci/scripts/go_build.sh /arrow &&
  /arrow/ci/scripts/go_test.sh /arrow"
```
or
```
$ archery docker info ubuntu-cuda-cpp --show environment
Service ubuntu-cuda-cpp docker compose config:
 - environment
  - ARROW_BUILD_STATIC: OFF
  - ARROW_CUDA: ON
  - ARROW_GANDIVA: OFF
  - ARROW_GCS: OFF
  - ARROW_ORC: OFF
  - ARROW_S3: OFF
  - ARROW_SUBSTRAIT: OFF
  - ARROW_WITH_OPENTELEMETRY: OFF
  - CCACHE_COMPILERCHECK: content
  - CCACHE_COMPRESS: 1
  - CCACHE_COMPRESSLEVEL: 6
  - CCACHE_MAXSIZE: 1G
  - CCACHE_DIR: /ccache

```
_edited only the commands to reflect the change of the command --show_

Lead-authored-by: Raúl Cumplido <raulcumplido@gmail.com>
Co-authored-by: Antoine Pitrou <pitrou@free.fr>
Signed-off-by: Antoine Pitrou <antoine@python.org>

thisisnic committed:
ARROW-18079: [R] Improve efficiency of schema creation to prevent performance regressions (#14447)

Authored-by: Nic Crane <thisisnic@gmail.com>
Signed-off-by: Nic Crane <thisisnic@gmail.com>

paleolimbot committed:
ARROW-17460: [R] Don't warn if the new UDF I'm registering is the same as the existing one (#14436)

This PR makes it so that you can do the following without a warning:

``` r
library(arrow, warn.conflicts = FALSE)

register_scalar_function(
  "times_32",
  function(context, x) x * 32L,
  in_type = list(int32(), int64(), float64()),
  out_type = function(in_types) in_types[[1]],
  auto_convert = TRUE
)

register_scalar_function(
  "times_32",
  function(context, x) x * 32L,
  in_type = list(int32(), int64(), float64()),
  out_type = function(in_types) in_types[[1]],
  auto_convert = TRUE
)
```

In fixing that, I also ran across an important discovery, which is that `cpp11::function` does *not* protect the underlying `SEXP` from garbage collection!!!! It the two functions we used this for were being protected by proxy because the execution environment of `register_scalar_function()` was being saved when the binding was registered.

Authored-by: Dewey Dunnington <dewey@voltrondata.com>
Signed-off-by: Dewey Dunnington <dewey@fishandwhistle.net>

fatemehp committed:
PARQUET-2179: [C++][Parquet] Add a test for skipping repeated fields (#14366)

Add a test for `TypedColumnReader::Skip` with repeated values to make it clear that we are skipping values and not records.
Also, add some comments to the existing test for Skip of non-repeated values.

Lead-authored-by: Fatemah Panahi <panahi@google.com>
Co-authored-by: Antoine Pitrou <antoine@python.org>
Co-authored-by: Fatemah Panahi <fatemehp@users.noreply.github.com>
Signed-off-by: Antoine Pitrou <antoine@python.org>

LouisClt committed:
ARROW-17524: [C++] Correction for fields included when reading an ORC table (#13962)

I think there is a bug in the ORC reader : when we specify the fields indexes that we want to keep, it does not work correctly. Looking at the code, it seems to be because we do "includeTypes" in lieue of "include" when setting the ORC options.
It can be problematic when we want to import an ORC table containing Union types as it will do an error at the import, even if we try not to import these specific fields.

The definitions of the corresponding ORC methods are here : 
https://github.com/apache/orc/blob/72220851cbde164a22706f8d47741fd1ad3db190/c%2B%2B/src/Options.hh#L185-L191

and 
https://github.com/apache/orc/blob/72220851cbde164a22706f8d47741fd1ad3db190/c%2B%2B/src/Options.hh#L201-L207

Lead-authored-by: LouisClt <louis1110@hotmail.fr>
Co-authored-by: Antoine Pitrou <antoine@python.org>
Co-authored-by: Antoine Pitrou <pitrou@free.fr>
Signed-off-by: Antoine Pitrou <antoine@python.org>

rtpsw commented:
@westonpace, I see this is a PR on your personal GitHub space. Is this intended?

westonpace commented:
> @westonpace, I see this is a PR on your personal GitHub space. Is this intended?

I think so.  My understanding is that we are going to do some rapid prototyping on this branch.

pitrou committed:
ARROW-18083: [C++] Bump vendored zlib version (#14446)

This new version includes a security fix.

Authored-by: Antoine Pitrou <antoine@python.org>
Signed-off-by: Sutou Kouhei <kou@clear-code.com>

nealrichardson committed:
ARROW-18080: [C++] Remove gcc <= 4.9 workarounds (#14441)

This is most of them I could find. There's another one in `python/pyarrow/src/arrow/python/datetime.cc` to remove but it's more involved.

Authored-by: Neal Richardson <neal.p.richardson@gmail.com>
Signed-off-by: Sutou Kouhei <kou@clear-code.com>

kou committed:
ARROW-18092: [CI][Conan] Update versions of gRPC related dependencies (#14453)

Error message:

https://github.com/ursacomputing/crossbow/actions/runs/3271941831/jobs/5382341820#step:5:566

    WARN: Remotes registry file missing, creating default one in /root/.conan/remotes.json
    WARN: grpc/1.48.0: requirement re2/20220601 overridden by arrow/10.0.0 to re2/20220201
    WARN: grpc/1.48.0: requirement protobuf/3.21.4 overridden by arrow/10.0.0 to protobuf/3.21.1
    WARN: googleapis/cci.20220711: requirement protobuf/3.21.4 overridden by grpc/1.48.0 to protobuf/3.21.1
    WARN: grpc-proto/cci.20220627: requirement protobuf/3.21.4 overridden by grpc/1.48.0 to protobuf/3.21.1
    ERROR: Missing binary: grpc/1.48.0:ddc600b3316e16c4e38f2c1ca1214d7241b4dd80
    grpc/1.48.0: WARN: Can't find a 'grpc/1.48.0' package for the
    specified settings, options and dependencies:
    - Settings: arch=x86_64,
                build_type=Release,
                compiler=gcc,
                compiler.libcxx=libstdc++,
                compiler.version=10,
                os=Linux
    - Options: codegen=True,
               cpp_plugin=True,
               csharp_ext=False,
               csharp_plugin=True,
               fPIC=True,
               node_plugin=True,
               objective_c_plugin=True,
               php_plugin=True,
               python_plugin=True,
               ruby_plugin=True,
               secure=False,
               shared=False,
               abseil:fPIC=True,
               abseil:shared=False,
               c-ares:fPIC=True,
               c-ares:shared=False,
               c-ares:tools=True,
               googleapis:fPIC=True,
               googleapis:shared=False,
               grpc-proto:fPIC=True,
               grpc-proto:shared=False,
               openssl:386=False,
               openssl:enable_weak_ssl_ciphers=False,
               openssl:fPIC=True,
               openssl:no_aria=False,
               openssl:no_asm=False,
               openssl:no_async=False,
               openssl:no_bf=False,
               openssl:no_blake2=False,
               openssl:no_camellia=False,
               openssl:no_cast=False,
               openssl:no_chacha=False,
               openssl:no_cms=False,
               openssl:no_comp=False,
               openssl:no_ct=False,
               openssl:no_deprecated=False,
               openssl:no_des=False,
               openssl:no_dgram=False,
               openssl:no_dh=False,
               openssl:no_dsa=False,
               openssl:no_dso=False,
               openssl:no_ec=False,
               openssl:no_ecdh=False,
               openssl:no_ecdsa=False,
               openssl:no_engine=False,
               openssl:no_filenames=False,
               openssl:no_gost=False,
               openssl:no_hmac=False,
               openssl:no_idea=False,
               openssl:no_md4=False,
               openssl:no_md5=False,
               openssl:no_mdc2=False,
               openssl:no_ocsp=False,
               openssl:no_pinshared=False,
               openssl:no_rc2=False,
               openssl:no_rfc3779=False,
               openssl:no_rmd160=False,
               openssl:no_rsa=False,
               openssl:no_seed=False,
               openssl:no_sha=False,
               openssl:no_sm2=False,
               openssl:no_sm3=False,
               openssl:no_sm4=False,
               openssl:no_sock=False,
               openssl:no_srp=False,
               openssl:no_srtp=False,
               openssl:no_sse2=False,
               openssl:no_ssl=False,
               openssl:no_ssl3=False,
               openssl:no_stdio=False,
               openssl:no_tests=False,
               openssl:no_threads=False,
               openssl:no_tls1=False,
               openssl:no_ts=False,
               openssl:no_whirlpool=False,
               openssl:openssldir=None,
               openssl:shared=False,
               protobuf:debug_suffix=True,
               protobuf:fPIC=True,
               protobuf:lite=False,
               protobuf:shared=False,
               protobuf:with_rtti=True,
               protobuf:with_zlib=True,
               re2:fPIC=True,
               re2:shared=False,
               zlib:fPIC=True,
               zlib:shared=False
    - Dependencies: abseil/20220623.0,
                    c-ares/1.18.1,
                    openssl/1.1.1q,
                    re2/20220201,
                    zlib/1.2.12,
                    protobuf/3.21.1,
                    googleapis/cci.20220711,
                    grpc-proto/cci.20220627
    - Requirements: abseil/20220623.Y.Z,
                    c-ares/1.Y.Z,
                    googleapis/cci.20220711,
                    grpc-proto/cci.20220627,
                    openssl/1.Y.Z,
                    protobuf/3.21.1:37dd8aae630726607d9d4108fefd2f59c8f7e9db,
                    re2/20220201.Y.Z,
                    zlib/1.Y.Z
    - Package ID: ddc600b3316e16c4e38f2c1ca1214d7241b4dd80

    ERROR: Missing prebuilt package for 'grpc/1.48.0'

Authored-by: Sutou Kouhei <kou@clear-code.com>
Signed-off-by: Sutou Kouhei <kou@clear-code.com>

kou committed:
ARROW-17353: [Release][R] Validate binaries version (#14396)

We generate version in Crossbow not CI job.
We can use file name validation feature by using version generated by Crossbow.

We use "X.Y.Z" version for RC version such as "10.0.0" for "10.0.0-rc1".

Authored-by: Sutou Kouhei <kou@clear-code.com>
Signed-off-by: Sutou Kouhei <kou@clear-code.com>

AlenkaF committed:
ARROW-17891: [Docs][Python] Update and sync Win section of the developers/python page (#14350)

This PR updates Windows section of the Python Development page. Main changes:

- use Python 3.10 (also in instructions for Linux/MacOs)
- definition of `PATH` not needed as Python doesn't search in `PATH` for dlls anymore ([3.8 +](https://bugs.python.org/issue43173))
- use `CONDA_PREFIX` to define `ARROW_HOME` as in other parts of the docs
- remove **Running C++ unit tests for Python integration** section (C++ unit tests are part of `pytest`-based test module as of https://github.com/apache/arrow/pull/14117)

cc @wjones127 @jorisvandenbossche 

Authored-by: Alenka Frim <frim.alenka@gmail.com>
Signed-off-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

jorisvandenbossche committed:
ARROW-17025:  [Dev] Remove github user name links from merge commit message (#14458)

Insert a space between the ``@`` and the username, to avoid it is an actual github reference

Authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>
Signed-off-by: Rok Mihevc <rok@mihevc.org>

lidavidm committed:
ARROW-17732: [Docs][Java] Add minimal JDBC driver docs (#14137)

Authored-by: David Li <li.davidm96@gmail.com>
Signed-off-by: David Li <li.davidm96@gmail.com>

kou committed:
ARROW-18103: [Packaging][deb][RPM] Fix upload artifacts patterns (#14462)

The current patterns may match multiple files that have the same base name. For example:

* `arrow/dev/tasks/linux-packages/apache-arrow/apt/repositories/debian/pool/bookworm/main/a/apache-arrow/libarrow-glib-dev_10.0.0.dev480-1_arm64.deb`
* `arrow/dev/tasks/linux-packages/apache-arrow/apt/build/debian-bookworm-arm64/libgandiva-glib-dev_10.0.0.dev480-1_arm64.deb`

The latter (`**/build/**`) is an artifact in a build directory. We should use only the former (`**/repositories/**`) that is an artifact for upload.


Authored-by: Sutou Kouhei <kou@clear-code.com>
Signed-off-by: Sutou Kouhei <kou@clear-code.com>

milesgranger committed:
ARROW-14596: [C++][Python] Read table nested struct fields in columns (#14326)

Part of [ARROW-14596](https://issues.apache.org/jira/browse/ARROW-14596), and [ARROW-13798](https://issues.apache.org/jira/browse/ARROW-13798)

Does not propose to solve selecting from lists in this PR; only supporting dotted paths into arbitrarily nested structs using existing `FromDotPath`. Selecting from lists will require further discussion and support for kernels to select struct subsets from lists.


Authored-by: Miles Granger <miles59923@gmail.com>
Signed-off-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

amoeba committed:
ARROW-15006: [Python][CI][Doc] Enable numpydoc check PR03 (#13983)

Adds an additional numypdoc check to CI (PR03) and fixes all corresponding violations. 

Note this does not fully resolve [ARROW-15006](https://issues.apache.org/jira/browse/ARROW-15006).

Authored-by: Bryce Mecum <petridish@gmail.com>
Signed-off-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

kou committed:
ARROW-17317: [Release][Docs] Normalize previous document version directory (#14457)

We should use X.Y instead of X.Y.Z (e.g.: 8.0 not 8.0.1) for previous version document directory.

See also:
https://github.com/apache/arrow/blob/apache-arrow-9.0.0/dev/release/post-08-docs.sh#L84

The script should accept X.Y.Z such as 8.0.1 and normalize it to X.Y. It'll reduce human error.

See also:

* https://github.com/apache/arrow-site/pull/228#issuecomment-1205997067
* https://github.com/apache/arrow-site/pull/228#issuecomment-1206085602

Authored-by: Sutou Kouhei <kou@clear-code.com>
Signed-off-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

amoeba committed:
MINOR: [R] [Docs] Fix link in ?cast documentation (#14455)

Lead-authored-by: Bryce Mecum <petridish@gmail.com>
Co-authored-by: Nic Crane <thisisnic@gmail.com>
Signed-off-by: Nic Crane <thisisnic@gmail.com>

amoeba committed:
MINOR: [R] Fix broken link in R developer setup vignette (#14443)

The link in the final section of the [Developer environment setup](https://arrow.apache.org/docs/dev/r/articles/developers/setup.html#other-installation-issues) docs 404s and this points the link to the installation details vignette. I think this is what was intended but let me know.

Authored-by: Bryce Mecum <petridish@gmail.com>
Signed-off-by: Nic Crane <thisisnic@gmail.com>

zeroshade committed:
ARROW-17871: [Go] initial binary arithmetic implementation (#14255)

Only implements Addition and Subtraction for integral types and float32/float64. Temporal types and others will come afterwards.

Authored-by: Matt Topol <zotthewizard@gmail.com>
Signed-off-by: Matt Topol <zotthewizard@gmail.com>

wjones127 committed:
ARROW-17954: [R] Update news for 10.0 (#14337)

Lead-authored-by: Will Jones <willjones127@gmail.com>
Co-authored-by: Neal Richardson <neal.p.richardson@gmail.com>
Co-authored-by: Dewey Dunnington <dewey@fishandwhistle.net>
Signed-off-by: Neal Richardson <neal.p.richardson@gmail.com>

rtpsw commented:
> I'll review the substrait changes more closely when we can figure out how to get rid of the `::substrait` change (I think Anja was making progress on this)?

I discussed this issue with @icexelloss. Instead of this `::substrait` change, a more localized change may be possible by a `using` statement for the namespace. I'll look into this.

zeroshade committed:
ARROW-18081: [Go] Add Scalar Boolean functions (#14442)

Authored-by: Matt Topol <zotthewizard@gmail.com>
Signed-off-by: Matt Topol <zotthewizard@gmail.com>

Y-- committed:
ARROW-18117: [C++] Fix static bundle build (#14465)

Which prevents `libarrow_bundled_dependencies` to contain all required symbols.

Authored-by: Y <y.lemaout@gmail.com>
Signed-off-by: Sutou Kouhei <kou@clear-code.com>

kou committed:
ARROW-18093: [CI][Conda][Windows] Disable ORC (#14454)

https://anaconda.org/conda-forge/orc doesn't provide binaries for Windows.

Error message:

https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=37759&view=logs&j=4c86bc1b-1091-5192-4404-c74dfaad23e7&t=41795ef0-6501-5db4-3ad4-33c0cf085626&l=497

    CMake Error at cmake_modules/FindORC.cmake:56 (message):
      ORC library was required in toolchain and unable to locate
    Call Stack (most recent call first):
      cmake_modules/ThirdpartyToolchain.cmake:280 (find_package)
      cmake_modules/ThirdpartyToolchain.cmake:4362 (resolve_dependency)
      CMakeLists.txt:496 (include)

Authored-by: Sutou Kouhei <kou@clear-code.com>
Signed-off-by: Sutou Kouhei <kou@clear-code.com>

jorisvandenbossche committed:
ARROW-18088: [CI][Python] Fix pandas master/nightly build failure related to timedelta (#14460)

Pandas 2.0+ will now preserve the unit of a timedelta64/datetime64 numpy array that is passed to a DataFrame constructor, instead of always coercing to nanoseconds. But pyarrow will still use nanoseconds in the conversion to pandas, giving data type mismatches in the tests. For now, ensuring we always create pandas.DataFrames with datetime64[ns] across pandas versions. Longer term solution is ARROW-18124

Authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>
Signed-off-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

zeroshade committed:
ARROW-17751: [Go][Benchmarking] Add Go Benchmark Script (#14148)

Authored-by: Matt Topol <zotthewizard@gmail.com>
Signed-off-by: Matt Topol <zotthewizard@gmail.com>

nealrichardson committed:
ARROW-18132: [R] Add deprecation cycle for pull() change (#14475)

Authored-by: Neal Richardson <neal.p.richardson@gmail.com>
Signed-off-by: Neal Richardson <neal.p.richardson@gmail.com>

nealrichardson committed:
ARROW-18132: [R] Add deprecation cycle for pull() change (#14475)

Authored-by: Neal Richardson <neal.p.richardson@gmail.com>
Signed-off-by: Neal Richardson <neal.p.richardson@gmail.com>

lidavidm committed:
MINOR: [Release] Add seccomp=undefined to APT verification (#14477)

Needed to avoid 'Failed to close file descriptor for child process'. The other Docker invocations already have it, but not this one.

Authored-by: David Li <li.davidm96@gmail.com>
Signed-off-by: Sutou Kouhei <kou@clear-code.com>

lidavidm committed:
MINOR: [Release] Fix loop in universal2 wheel verification (#14479)

Authored-by: David Li <li.davidm96@gmail.com>
Signed-off-by: Sutou Kouhei <kou@clear-code.com>

rtpsw committed:
try smaller fix of MSVC namespace errors

rtpsw committed:
second try

rtpsw committed:
third try

rtpsw committed:
requested changes

rtpsw committed:
Merge branch 'feature/bamboo-demo' into ARROW-17980

rtpsw committed:
fix serde test

rtpsw committed:
Merge branch 'master' into ARROW-17980
